### PR TITLE
Finished Part2

### PR DESCRIPTION
**Pull request by jaysomething**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+struct IntType;
+
 struct FloatType
 {
-    float add(float lhs, float rhs);
-    float subtract(float lhs, float rhs);
-    float multiply(float lhs, float rhs);
-    float divide(float lhs, float rhs);
+    FloatType() {}
+    FloatType(float val)
```
Comment (matkatmusic) [inline main.cpp]: You can simplify this via the member initializer list
```
FloatType(float val) : value( new float(val) ) { }
```

File: main.cpp [hunk 2]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+struct DoubleType;
+struct IntType;
+
 struct FloatType
 {
-    float add(float lhs, float rhs);
-    float subtract(float lhs, float rhs);
-    float multiply(float lhs, float rhs);
-    float divide(float lhs, float rhs);
+    FloatType() {}
```
Comment (matkatmusic) [inline main.cpp]: You should not need this default constructor

- [x] remove default constructor

File: main.cpp [hunk 3]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+};
+
+struct DoubleType
+{
+    DoubleType() {}
+    DoubleType(double val) 
+    { 
+        value = new double;
+        *value = val;
+    }
```
Comment (matkatmusic) [inline main.cpp]: Same comments in FloatType apply here.

File: main.cpp [hunk 4]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+};
+
+struct IntType
+{
+    IntType() {}
+    IntType(int val) 
+    { 
+        value = new int;
+        *value = val;
+    }
```
Comment (matkatmusic) [inline main.cpp]: Same comments in FloatType apply here

File: main.cpp [hunk 5]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    std::cout << *value << '\n';
+    return *this;
+}
+
+FloatType& FloatType::divide(const FloatType& ft)
+{
+    if (*ft.value == 0)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
```
Comment (matkatmusic) [inline main.cpp]: floating point division-by-zero is well-defined, and returns `inf`.  It's not an error.  it's only an error for `int` division-by-zero

- [x] correct warning

File: main.cpp [hunk 6]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    std::cout << *value << '\n';
+    return *this;
+}
+
+FloatType& FloatType::divide(const IntType& it)
+{
+    if (*it.value == 0)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
```
Comment (matkatmusic) [inline main.cpp]: correct the warning

File: main.cpp [hunk 7]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    std::cout << *value << '\n';
+    return *this;
+}
+
+DoubleType& DoubleType::divide(double v)
+{
+    if (v == 0.0)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
```
Comment (matkatmusic) [inline main.cpp]: correct the warning

File: main.cpp [hunk 8]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    std::cout << *value << '\n';
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const FloatType& ft)
+{
+    if (*ft.value == 0)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
```
Comment (matkatmusic) [inline main.cpp]: correct the warning

File: main.cpp [hunk 9]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
 
-int IntType::divide(int lhs, int rhs)
+DoubleType& DoubleType::divide(const DoubleType& dt)
 {
-    if (rhs == 0)
+    if (*dt.value == 0)
     {
-        std::cout << "Error: cannot divide by 0! Returning 0 instead\n";
-        return 0;
+        std::cout << "Error: cannot divide by 0!\n";
```
Comment (matkatmusic) [inline main.cpp]: correct the warning

File: main.cpp [hunk 10]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    std::cout << *value << '\n';
+    return *this;
+}
+
+DoubleType& DoubleType::divide(const IntType& it)
+{
+    if (*it.value == 0)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
```
Comment (matkatmusic) [inline main.cpp]: correct the warning

File: main.cpp [hunk 11]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    std::cout << *value << '\n';
+    return *this;
+}
+
+IntType& IntType::divide(int v)
+{
+    if (v == 0.0f)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
```
Comment (matkatmusic) [inline main.cpp]: `v` is an int, yet you're comparing it with a `float`

You also didn't prevent the division from happening.

File: main.cpp [hunk 12]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    std::cout << "(it) " << *value << " / " << v << " = ";
+    *value /= v;
```
Comment (matkatmusic) [inline main.cpp]: - [x] prevent division-by-zero if `v == 0`

File: main.cpp [hunk 13]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+
+IntType& IntType::divide(const FloatType& ft)
+{
+    if (*ft.value == 0)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
+    std::cout << "(it) " << *value << " / (ft) " << *ft.value << " = ";
+    *value = *value / *ft.value;
+    std::cout << *value << '\n';
```
Comment (matkatmusic) [inline main.cpp]: I could be mistaken, but I believe you might get warnings about this line and the implicit conversion.   You'll have to try feeding a `FloatType(0)` to this function and see if you get a crash.

File: main.cpp [hunk 14]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+
+IntType& IntType::divide(const DoubleType& dt)
+{
+    if (*dt.value == 0)
+    {
+        std::cout << "Error: cannot divide by 0!\n";
+    }
+    std::cout << "(it) " << *value << " / (dt) " << *dt.value << " = ";
+    *value = *value / *dt.value;
+    std::cout << *value << '\n';
```
Comment (matkatmusic) [inline main.cpp]: Same issue as with FloatType being passed in mentioned above.

File: main.cpp [hunk 15]
```diff
@@ -64,151 +63,592 @@ New/This/Pointers/References conclusion
...
+    return *this;
+}
+
+IntType& IntType::divide(const IntType& it)
+{
+    if (*it.value == 0)
+    {
+        std::cout << "Error: cannot divide by 0! Returning zero instead\n";
+        *value = 0;
```
Comment (matkatmusic) [inline main.cpp]: Why even change `*value` though?

- [x] rethink this implementation

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: main.cpp (hunk 13 above)
Comment (jaysomething) [reply main.cpp]: This is the result:
`(it) 1 / (ft) 0 = -2147483648`
I'll skip the calculation in this case as well.

I have suppressed 3 warning flags in .replit though:
`-Wno-float-equal -Wno-conversion -Wno-double-promotion`
Am i actually not supposed to have done this?

File: main.cpp (hunk 14 above)
Comment (jaysomething) [reply main.cpp]: Same result as above
`(it) 1 / (dt) 0 = -2147483648`

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**